src/components/DefaultPageLayout.tsx

```diff
@@ -85,60 +97,71 @@ const DefaultPageLayout = ({
     );
   }, [intl, navigate, toast]);
 
+  const ProjectDropTargetWhenNeeded =
+    isTrainDialogClosed &&
+    isTourClosed &&
+    isConnectionDialogClosed &&
+    isSaveDialogClosed
+      ? ProjectDropTarget
+      : React.Fragment;
+
   return (
     <>
       {/* Suppress dialogs to prevent overlapping dialogs */}
-      {!isEditorOpen && stage === TrainModelDialogStage.Closed && (
-        <ConnectionDialogs />
-      )}
+      {!isEditorOpen &&
+        isTrainDialogClosed &&
+        isTourClosed &&
+        isSaveDialogClosed && <ConnectionDialogs />}
       {!isEditorOpen && <TrainModelDialogs />}
-      {!isEditorOpen && stage === TrainModelDialogStage.Closed && <Tour />}
+      {!isEditorOpen && <Tour />}
```

Comment:
This approach kinda feels like it's reached its limit. One to revisit when we've consolidated the state.

Reply:
Agreed. That would be quite satisfying to consolidate

Reply:
Was it not viable to move the project drop overlay to a z-index below the modal dialog overlay?

Reply:
Haven't tried that. Will give that a go

Reply:
Seems like the part detecting the drag over doesn't have a zIndex (https://github.com/microbit-foundation/ml-trainer/blob/drop-target-when-needed/src/components/FileDropTarget.tsx#L47).

I also tried changing the z-Index of the project drop overlay. The onDragOver is still triggered, whilst the  onDragLeave does not get triggered. The result is that the overlay is stuck on screen.